### PR TITLE
POC: refactor switch with hookz

### PR DESCRIPTION
**Pull request by whatknight**
This is a first pass refactoring a Switcher with hooks. **Not ready to merge.**

## Notables
- had to swap out enzyme for react-testing-library since it doesn't support out of band hooks like `useEffect`, this led to some workarounds to test api boundaries and contracts.
- ~cWRP case isn't tested so it didn't make the refactor, we need to add a test and then probably add each one of the listener props to second arg of `useEffect`.~ This has been refactored. The items in state are now calculated on each render. This was defacto happening before anyways since the call was happening in cWRP. Since there's no more state to update, we are forcing an update on state changes.
- ~`preventUpdate` prop is not supported, though I'd argue its never necessary and would just cause bugs.~ this has been implemented for backwards compat, but I still think it should be removed in a major release-- maybe add a dev warning for deprecation?
- ~theres a message when you run the test that complains about a state change being fired outside of act. I think this is a false  positive caused by the component for the testing the context provider due to legacy context being defined in the test file, but I haven't tried to move it out yet.~ `act` doesn't work with async await. Using fake timers removed the need for async await. 
- had to add the transform-runtime to get `async/await` to work in tests, though i could have added too much in my quest to run the tests. _update: this might be unnecessary now_
- the Switcher test def needs a second pass since I was figuring out hacks to test stuff as I went along.

Have you had any luck adding https://www.npmjs.com/package/rollup-plugin-serve or something to be able to test this lib in dev?

### CODE REVIEW

**Comment by jdlehman** [comment]
@whatknight what is left to do in this work? if it's ready I'll do a thorough review.

**Comment by whatknight** [comment]
I'd say the last major todo is to handle the prop updates that the cWRP was there for, but I don't know if its necessary before a review pass. I started down that road but writing the test was a little tricky and  I ran out of time.

**Comment by whatknight** [comment]
added the new lint rule to catch deps of hooks. and led me down a refactoring path which made me feel like i got hooks a little better. still need to do the tests for the prop changes.

**Comment by whatknight** [comment]
this is ready for review.

File: package.json [hunk 1]
```diff
@@ -55,25 +55,27 @@
...
     "husky": "^1.3.1",
     "jest": "^24.1.0",
+    "jest-dom": "^3.1.1",
     "jsdom": "^13.2.0",
     "lint-staged": "^8.1.3",
     "prettier": "^1.16.4",
-    "react": "^16.8.1",
-    "react-dom": "^16.8.1",
-    "react-test-renderer": "^16.8.1",
+    "react": "^16.8.3",
```
Comment (jdlehman) [inline package.json]: We should also update the peer dependency requirement to make it ^16.8.0

File: src/Switcher.js [hunk 2]
```diff
@@ -8,185 +9,188 @@ import {
...
-    const visibleSwitch = getSwitch(path, props);
-    const activePath = getActivePath(path, props.basePath, visibleSwitch);
-    const dynamicValues = getDynamicSegments(
-      path,
-      props.basePath,
-      visibleSwitch
-    );
+    const id = generateGuid();
+    load &&
+      switcherProvider.loadListeners.push({
```
[the remark this answers is not in the document]
Comment (jdlehman) [reply src/Switcher.js]: For reference: https://reactjs.org/docs/context.html#updating-context-from-a-nested-component

**Pull request merged**